### PR TITLE
 Add more Pen tests, fix bugs and cleanup the source

### PR DESCRIPTION
**Pull request by hughbe**
Fixes https://github.com/mono/libgdiplus/issues/473

### CODE REVIEW

File: src/pen.c [hunk 1]
```diff
@@ -252,31 +252,26 @@ GdipCreatePen1 (ARGB argb, REAL width, GpUnit unit, GpPen **pen)
...
 		return GdiplusNotInitialized;
 
-	if (!pen || unit < UnitWorld || unit > UnitCairoPoint || unit == UnitDisplay)
+	if (!pen || unit > UnitCairoPoint || unit == UnitDisplay)
 		return InvalidParameter;
 
 	result = gdip_pen_new ();
 	if (!result)
-		goto error;
+		return OutOfMemory;
```
Comment (filipnavara) [inline src/pen.c]: This leaves *pen uninitialized, which is not a good practice.

File: src/pen.c [hunk 2]
```diff
@@ -252,31 +252,26 @@ GdipCreatePen1 (ARGB argb, REAL width, GpUnit unit, GpPen **pen)
...
 	result->color = argb;
 	result->width = width;
 	result->unit = unit;
 	result->own_brush = TRUE;
 
 	status = GdipCreateSolidFill (argb, (GpSolidFill **) &result->brush);
-	if (status != Ok)
-		goto error;
+	if (status != Ok) {
+		GdipDeletePen (result);
```
Comment (filipnavara) [inline src/pen.c]: Same here.

Comment (filipnavara) [reply src/pen.c]: ...and in GdipCreatePen2 below as well.

**Comment by hughbe** [comment]
Well actually I think Windows would crash in these situations (but we can't actually reach it realistically because these allocations wouldn't really fail normally). So we're free to do whatever we want, so I added these assignments back

**Pull request merged**